### PR TITLE
[SAP] Create way for user to create password

### PR DESCRIPTION
**Pull request by polybuildr**
Replaces #88 (cleaner commits). @hharchani needs to clean up his local repo though. :P Also, he'll probably be busy today, so we can continue work here, @nisargjhaveri.

### CODE REVIEW

**Comment by polybuildr** [comment]
Comments at https://github.com/hharchani/felicity16-website/commit/ec67c4a1a3965321784bf8a516103801963d5fce are still completely relevant.

File: src/app/controllers/sap.php [hunk 1]
```diff
@@ -68,6 +68,53 @@ public function logout() {
...
 
+    public function verify($hash="") {
+        $this->load_library('http_lib');
+        if ($hash == "") {
+            $this->http_lib->response_code(404);
+        }
+        $this->load_model('sap_model');
+        $user = $this->sap_model->verify_hash($hash);
+        if (! $user) {
+            $this->http_lib->response_code(404);
```
Comment (nisargjhaveri) [inline src/app/controllers/sap.php:79]: Isn't there any http status for improper and incomplete data?

Comment (polybuildr) [inline src/app/controllers/sap.php:79]: I skimmed through the https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_Client_Error. There seem to be some that sound right, but they're about HTTP headers, or in general, don't match the kind of case we're dealing with here. :/ I might be wrong, but can't find one right now.

File: src/app/controllers/sap.php [hunk 2]
```diff
@@ -68,6 +68,53 @@ public function logout() {
...
+        }
+        $this->load_library('csrf_lib');
+        $this->load_library('session_lib');
+        $this->load_view('sap/create_password', [
+            'csrf_token' => $this->csrf_lib->new_csrf_token(),
+            'hash' => $hash,
+            'error' => $this->session_lib->flash_get('error'),
```
Comment (nisargjhaveri) [inline src/app/controllers/sap.php:86]: Where does this gets set? This link is to be sent in the email, right?

Comment (polybuildr) [inline src/app/controllers/sap.php:86]: Yeah, this is the link that goes in the email. The error gets set in line 113, when the user tries to set a password, but the application fails to set it.

File: src/app/controllers/sap_portal.php [hunk 3]
```diff
@@ -214,4 +214,87 @@ private function validate_data($posted_data, $rules) {
...
+        ob_start();
+        $this->load_view('sap/create_password_email_text', [
+            'subject' => $subject,
+            'name' => $user['name'],
+            'link' => $link
+        ]);
+        $email_body_text = ob_get_contents();
+        ob_end_clean();
+
+        $this->email->send_mail([
```
Comment (nisargjhaveri) [inline src/app/controllers/sap_portal.php:270]: We need to update email lib, it is soon gonna start having problems..

Comment (polybuildr) [inline src/app/controllers/sap_portal.php:270]: Sure. But of what kind? :P

File: src/app/models/sap_model.php [hunk 4]
```diff
@@ -284,4 +284,88 @@ public function submit_review($submission_id, $approved) {
             false
         );
     }
+
+    public function get_users_list() {
+        $stmt = $this->DB->sap->query('SELECT * FROM `sap_ambassadors` WHERE `is_removed` = 0');
```
Comment (nisargjhaveri) [inline src/app/models/sap_model.php]: Well, ideally it is not stmt but a result. Just saying.

File: src/app/models/sap_model.php [hunk 5]
```diff
@@ -284,4 +284,88 @@ public function submit_review($submission_id, $approved) {
...
+            $this->DB->sap,
+            'UPDATE `sap_ambassadors` SET `hash_for_ceating_password`=? WHERE `id`=?',
+            'si',
+            [$password_hash, $id],
+            false
+        );
+    }
+
+    public function remove_user($id) {
+        $this->db_lib->prepared_execute(
```
Comment (nisargjhaveri) [inline src/app/models/sap_model.php]: A return would be nice to have.

File: src/app/models/sap_model.php (hunk 4 above)
Comment (polybuildr) [inline src/app/models/sap_model.php]: You're right, though. :P I'll change the name.

File: sql/migrations/0003_add_sap_password_creation.sql [hunk 6]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE `sap_ambassadors`
+ADD `hash_for_ceating_password` CHAR(42) NULL DEFAULT NULL UNIQUE AFTER `registration_time`,
+ADD `has_activated` BOOLEAN NOT NULL DEFAULT FALSE AFTER `hash_for_ceating_password`,
```
Comment (nisargjhaveri) [inline sql/migrations/0003_add_sap_password_creation.sql:3]: Not `is_activated`?

File: src/app/models/sap_model.php (hunk 5 above)
Comment (polybuildr) [inline src/app/models/sap_model.php]: Oops. Not just nice to have, required! I'm using the return value in the controller. :P Fixing.

Comment (nisargjhaveri) [inline src/app/models/sap_model.php]: o.O I didn't see that. :P

File: src/app/views/sap/create_password.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+            <label>Password: <input name="password" type="password" required></label>
+            <label>Confirm Password: <input name="repassword" type="password" required></label>
+            <p class="text-error password-error"></p>
+            <input type="hidden" name="csrf_token" value="<?= $csrf_token ?>">
+            <input type="hidden" name="hash" value="<?= $hash ?>">
+            <input type="submit" value="Submit">
+        </form>
+        <script type="text/javascript">
+            window.onload = function () {
+                document.querySelector('form').onsubmit = function (e) {
```
Comment (nisargjhaveri) [inline src/app/views/sap/create_password.php]: Harshit uses a lots of query selectors, doesn't he? :P

And, better to have onsubmit directly on the form. Too generic selector. and this wastes time onload.

File: src/app/views/sap/create_password.php [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+                    if (this.password.value !== this.repassword.value) {
+                        e.preventDefault();
```
Comment (nisargjhaveri) [inline src/app/views/sap/create_password.php:28]: Careful, `e` will have some problems with onload.

**Comment by nisargjhaveri** [comment]
Scrutinizer has some style suggestions, if you want to have a look. Ignore the issues.

**Comment by nisargjhaveri** [comment]
Otherwise, looks fine. Squash?

**Comment by polybuildr** [comment]
Fixed styling issues. I'll squash and merge. I wish there was a way to keep these commits separately somewhere, but keep the squashed version in the main repo.

**Pull request merged**